azure/services/natgateways/spec.go

```diff
@@ -0,0 +1,105 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package natgateways
+
+import (
+	"fmt"
+
+	"github.com/Azure/azure-sdk-for-go/services/network/mgmt/2021-02-01/network"
+	autorest "github.com/Azure/go-autorest/autorest/azure"
+	"github.com/Azure/go-autorest/autorest/to"
+	"github.com/pkg/errors"
+	infrav1 "sigs.k8s.io/cluster-api-provider-azure/api/v1beta1"
+	"sigs.k8s.io/cluster-api-provider-azure/azure"
+)
+
+// NatGatewaySpec defines the specification for a Nat Gateway.
+type NatGatewaySpec struct {
+	Name           string
+	ResourceGroup  string
+	SubscriptionID string
+	Location       string
+	SubnetName     string
+	NatGatewayIP   infrav1.PublicIPSpec
+}
+
+// ResourceName returns the name of the Nat Gateway.
+func (s *NatGatewaySpec) ResourceName() string {
+	return s.Name
+}
+
+// ResourceGroupName returns the name of the resource group.
+func (s *NatGatewaySpec) ResourceGroupName() string {
+	return s.ResourceGroup
+}
+
+// OwnerResourceName returns the name of the associated subnet.
+func (s *NatGatewaySpec) OwnerResourceName() string {
+	return s.SubnetName
+}
+
+// Parameters returns the parameters for the Nat Gateway.
+func (s *NatGatewaySpec) Parameters(existing interface{}) (interface{}, error) {
+	if existing != nil {
+		existingNatGateway, ok := existing.(network.NatGateway)
+		if !ok {
+			return nil, errors.Errorf("%T is not a network.NatGateway", existing)
+		}
+		// Nat Gateway already exists, parse existing spec
+		publicIP, err := GetPublicIP(existingNatGateway)
+		if err != nil {
+			return nil, err
+		}
+		if s.NatGatewayIP.Name == publicIP {
+			// Skip update for Nat Gateway as it exists with expected values
+			return nil, nil
+		}
+	}
+
+	natGatewayToCreate := network.NatGateway{
+		Location: to.StringPtr(s.Location),
+		Sku:      &network.NatGatewaySku{Name: network.NatGatewaySkuNameStandard},
+		NatGatewayPropertiesFormat: &network.NatGatewayPropertiesFormat{
+			PublicIPAddresses: &[]network.SubResource{
+				{
+					ID: to.StringPtr(azure.PublicIPID(s.SubscriptionID, s.ResourceGroupName(), s.NatGatewayIP.Name)),
+				},
+			},
+		},
+	}
+
+	return natGatewayToCreate, nil
+}
+
+func GetPublicIP(natGateway network.NatGateway) (string, error) {
```

Comment:
an idea: we could change this to `hasPublicIP` (don't think it needs to be public, it's just used here) and pass in the publicIP name, and just return true if the nat gateway has that public IP in its `PublicIPAddresses`